lib/faye/websocket/client.rb

```diff
@@ -62,13 +65,24 @@ def on_connect(stream)
 
         if @secure
           socket_tls = {:sni_hostname => URI.parse(@url).host}.merge(@socket_tls)
+          add_trust_ca(socket_tls.delete(:trust_ca))
           @stream.start_tls(socket_tls)
         end
 
         worker = @proxy || @driver
         worker.start
       end
 
+      def add_trust_ca(ca_file)
+        return if ca_file.nil?
+        @trust_ca = Array[ca_file].map{|ca| OpenSSL::X509::Certificate.new(File.read(ca)) }
+      end
+
+      def ssl_verify_peer(cert)
+        crt = OpenSSL::X509::Certificate.new(cert)
+        return @cert_store.verify(crt) || @trust_ca.any?{|ca| ca.verify(crt.public_key) }
```

Comment:
I didn't test this code yet, but immediate problems that I see with this:
* The EM OpenSSL verify callback ignores the preverify errors, i.e. it doesn't care if the server-supplied cert and private key do not match: https://github.com/eventmachine/eventmachine/issues/275 https://github.com/eventmachine/eventmachine/blob/v1.2.3/ext/ssl.cpp#L589
* This callback gets called with each individual cert in the chain that the server sends? How will it work if the server sends an intermediate CA cert that's not in the cert store?
* This doesn't actually validate the depth=0 cert's CN against the URI host, so it will accept any valid cert for any domain?
* NotBefore/NotAfter checks etc